catalyst_voices_packages/catalyst_cardano_serialization/lib/src/witness.dart

```diff
@@ -1,35 +1,161 @@
   final Set<VkeyWitness> vkeyWitnesses;
 
+  /// The native scripts.
+  final Set<NativeScript> nativeScripts;
+
+  /// The Plutus V1 scripts.
+  final Set<PlutusV1Script> plutusV1Scripts;
+
+  /// Redeemers for Plutus scripts.
+  // FIXME(ilap): implement proper redeemers type.
+  final Set<Redeemer> redeemers;
+
+  /// The Plutus V2 scripts.
+  final Set<PlutusV2Script> plutusV2Scripts;
+
+  /// The Plutus V3 scripts.
+  final Set<PlutusV3Script> plutusV3Scripts;
+
   /// The default constructor for [TransactionWitnessSet].
-  const TransactionWitnessSet({required this.vkeyWitnesses});
+  const TransactionWitnessSet({
+    this.vkeyWitnesses = const {},
+    this.nativeScripts = const {},
+    this.plutusV1Scripts = const {},
+    this.redeemers = const {},
+    this.plutusV2Scripts = const {},
+    this.plutusV3Scripts = const {},
+  });
 
   /// Deserializes the type from cbor.
   factory TransactionWitnessSet.fromCbor(CborValue value) {
     final map = value as CborMap;
     return TransactionWitnessSet(
-      vkeyWitnesses: map.values.map(VkeyWitness.fromCbor).toSet(),
+      vkeyWitnesses: _getWitnesses<VkeyWitness>(
+        map,
+        WitnessType.vkeyWitness,
+        VkeyWitness.fromCbor,
+      ),
+      nativeScripts: _getWitnesses<NativeScript>(
+        map,
+        WitnessType.nativeScript,
+        NativeScript.fromCbor,
+      ),
+      plutusV1Scripts: _getWitnesses<PlutusV1Script>(
+        map,
+        WitnessType.plutusV1Script,
+        PlutusV1Script.fromCbor,
+      ),
+      redeemers: _getWitnesses<Redeemer>(
+        map,
+        WitnessType.redeemers,
+        Redeemer.fromCbor,
+      ),
+      plutusV2Scripts: _getWitnesses<PlutusV2Script>(
+        map,
+        WitnessType.plutusV2Script,
+        PlutusV2Script.fromCbor,
+      ),
+      plutusV3Scripts: _getWitnesses<PlutusV3Script>(
+        map,
+        WitnessType.plutusV3Script,
+        PlutusV3Script.fromCbor,
+      ),
     );
   }
 
+  static Set<T> _getWitnesses<T>(
+    CborMap map,
+    WitnessType type,
+    T Function(CborValue) fromCbor,
+  ) {
+    final value = map[CborSmallInt(type.value)];
+    return value != null ? (value as CborList).map(fromCbor).toSet() : {};
+  }
+
   /// Serializes the type as cbor.
   @override
   CborValue toCbor() {
     return CborMap({
-      for (final vkey in vkeyWitnesses.indexed)
-        CborSmallInt(vkey.$1): vkey.$2.toCbor(),
+      ..._generateCborPair(WitnessType.vkeyWitness, vkeyWitnesses),
+      ..._generateCborPair(WitnessType.nativeScript, nativeScripts),
+      ..._generateCborPair(WitnessType.plutusV1Script, plutusV1Scripts),
+      ..._generateCborPair(WitnessType.redeemers, redeemers),
+      ..._generateCborPair(WitnessType.plutusV2Script, plutusV2Scripts),
+      ..._generateCborPair(WitnessType.plutusV3Script, plutusV3Scripts),
     });
   }
 
+  Map<CborValue, CborValue> _generateCborPair(
+    WitnessType witnessType,
+    Set<dynamic> witnesses,
```

Comment:
Imho we can pass here set of CborEncodable's, then we can avoid the type casting in the body.
```suggestion
    Set<CborEncodable> witnesses,
```

Reply:
Yeah, missed it.